### ISSUE TITLE
garbage-collect apps in client

### ISSUE THREAD

**Issue by davidpanderson** [BOINC/boinc#2212]
Currently the client doesn't garbage-collect apps, so files for old apps remain on client.

How to do this:
if sched reply includes any apps, mark other apps as deprecated.
In garbage collection, if an app is deprecated and not referenced by any WUs,
delete it and all APP_VERSIONs that refer to it.
Do this before garbage-collecting files

**Comment by JuhaSointusalo** [comment]
sched_reply includes only those apps that the scheduler assigned work for on that sched_reply. The client may occasionally run out of work for some particular app when project has several apps. So I think that in your scheme apps will be garbage collected too early, when the app isn't 'old' yet.

How about doing it like this instead:

- add a last used attribute to app version
- update the attribute every time a task finishes
- if app version is last used over a year ago and isn't referenced by any task garbage collect it
- if an app isn't referenced by any app version garbage collect it

The last used is an attribute of app version so that no longer used plan classes can be garbage collected even when the app is otherwise still active.

**Comment by Ageless93** [comment]
I'm also wondering whether that shouldn't be the sched_request file, as the sched_reply can be without mention of any apps, and only mention the PNG files. 
My present sched_reply for Seti mentions tasks and the PNG files, but no apps. 
My sched_request file for Seti shows the contents of my app_info.xml anonymous platform file, including all apps mentioned therein.

**Comment by davidpanderson** [comment]
True.  I suggest that we change the scheduler reply to return all non-deprecated apps.

**Comment by JuhaSointusalo** [comment]
> I suggest that we change the scheduler reply to return all non-deprecated apps.

Pointing out the obvious: that would require some new structures in the reply so that the client can tell apart a sched_reply that includes only one app because the scheduler is old and a sched_reply that includes only one app because that's all the project has.

Returning all non-deprecated apps in the reply feels a bit silly to me considering that 99,99% of time there is no change, the scheduler returns the same apps it returned the previous 10000 times.

`apps.php` lists non-deprecated apps. We could have the client download the list every once in a while.

**Comment by davidpanderson** [comment]
Given a tradeoff between (slightly) increased scheduler reply and increased code complexity, I'd take the former 100% of the time.

**Comment by Ageless93** [comment]
This is a duplicate of #1386
